python/tskit/tables.py

```diff
@@ -4196,6 +4199,142 @@ def union(
+
+        .. seealso::
+            :meth:`.union` for a combining two tables nodewise
+
+        :param TableCollection other: Another table collection.
+        :param list node_mapping: An array of node IDs that relate nodes in
+            ``other`` to nodes in ``self``: the k-th element of ``node_mapping``
+            should be the index of the equivalent node in ``self``, or
+            :data:`tskit.NULL` if the node is not present in ``self`` (in which
+            case it will be added to ``self``).
+        :param bool record_provenance: If ``True`` (default), record details of this
+            call to ``merge`` in the returned tree sequence's provenance
+            information (Default: ``True``).
+        :param bool add_populations: If ``True`` (default), populations referred
+            to from nodes new to ``self`` will be added as a new populations.
+            If ``False`` new populations will not be created, and populations
+            with the same ID in ``self`` and the other tree sequences will be reused.
+        :param bool check_populations: If ``True`` (default), check that the
+            populations referred to from nodes in the other tree sequences
+            are identical to those in ``self``.
+        :raises ValueError: If the node mapping is not of the correct length,
+            or if the sequence lengths of the two tree sequences are not equal,
+            or if the populations referred to from nodes in the other tree sequence
+            do not match those in ``self``.
+        """
+        if add_populations is None:
+            add_populations = True
+        if check_populations is None:
+            check_populations = True
+
+        node_mapping = util.safe_np_int_cast(node_mapping, np.int32)
+        node_map = node_mapping.copy()
+        if node_map.shape != (other.nodes.num_rows,):
+            raise ValueError(
+                "node_mapping must be of length equal to the number of nodes in other"
+            )
+        if self.sequence_length != other.sequence_length:
+            raise ValueError(
+                "Tree sequences must have same sequence lengths: use trim or shift to"
+                " adjust sequence lengths as necessary"
+            )
+        if check_populations:
+            nodes_pop = other.nodes.population
+            if add_populations:
+                # Only need to check those populations used by nodes in the node_mapping
+                nodes_pop = nodes_pop[node_map != tskit.NULL]
+            for i in np.unique(nodes_pop[nodes_pop != tskit.NULL]):
+                if (
+                    i >= self.populations.num_rows
+                    or self.populations[i] != other.populations[i]
+                ):
+                    raise ValueError("Non-matching populations")
+        individual_map = {}
+        population_map = {}
+        for new_node in np.where(node_map == tskit.NULL)[0]:
```

Comment:
Perhaps we should check that the nodes in the node map are identical? So something like
```
for oj, sj in enumerate(node_map):
    if sj != tskit.NULL:
        if self.node(sj) != other.node(oj):
            raise ValueError("Nodes do not match...")
    else:
          # continue as below
```
```